spec/mongoid/criteria_spec.rb

```diff
@@ -3422,20 +3422,139 @@ def self.ages; self; end
 
       context "when querying on a big decimal" do
 
-        let(:sales) do
-          BigDecimal('0.1')
+        context 'when map_big_decimal_to_decimal128 is false' do
+
+          around do |example|
+            current = Mongoid.map_big_decimal_to_decimal128
+            Mongoid.map_big_decimal_to_decimal128 = false
+            example.run
+            Mongoid.map_big_decimal_to_decimal128 = current
+          end
+
+          let(:sales) do
+            BigDecimal('0.1')
+          end
+
+          let!(:band) do
+            Band.create!(name: "Boards of Canada", sales: sales)
+          end
+
+          let(:from_db) do
+            Band.where(sales: sales).first
+          end
+
+          it "finds the document by the big decimal value" do
+            expect(from_db).to eq(band)
+          end
         end
 
-        let!(:band) do
-          Band.create!(name: "Boards of Canada", sales: sales)
+        context 'when map_big_decimal_to_decimal128 is true' do
+
+          around do |example|
+            current = Mongoid.map_big_decimal_to_decimal128
+            Mongoid.map_big_decimal_to_decimal128 = true
+            example.run
+            Mongoid.map_big_decimal_to_decimal128 = current
+          end
+
+          let(:sales) do
+            BigDecimal('0.1')
+          end
+
+          let!(:band) do
+            Band.create!(name: "Boards of Canada", sales: sales)
+          end
+
+          let(:from_db) do
+            Band.where(sales: sales).first
+          end
+
+          it "finds the document by the big decimal value" do
+            expect(from_db).to eq(band)
+          end
         end
 
-        let(:from_db) do
-          Band.where(sales: sales).first
+        context 'when map_big_decimal_to_decimal128 was false and is now true' do
```

Comment:
This test does not work. It asserts that a BigDecimal that was originally stored as a string, can be queried with a BigDecimal, with `map_big_decimal_to_decimal128` turned on. We have to fix this. Here are the solutions I can think of:
1. Change the implementation of searching for a `BigDecimal` field to search for both the string representation and the BigDecimal representation.
A problem arises when some people purposely have `BigDecimal`s and Strings stored in the database under the same field, and don't want querying for a `BigDecimal` to return the string. But I think this is a pretty niche case.
2. We can just leave it how it is. Instruct the user how to query for both string and `BigDecimal` representations of their data if that's what they want. This is a little annoying for the user.
3. I also wrote up this quick little query that changes all values in a db to a ``BigDecimal`` but this seems a little reckless, especially for big data sets:
```
db.bands.updateMany({ 
  "field": { "$exists": true } 
}, [
  { "$set": { 
    "field": { "$toDecimal": "$field" } 
  } }
])
```

Reply:
The first idea would be a really good one but I expect it to be either difficult or perhaps not even possible.

There are two contexts where attribute values are used: implicit equality (e.g. `{field: 42}`) and operators (e.g. `{field: {$eq: 42}}`). While the implicit equality can be transformed to e.g. `{field: {$in: ['42', Decimal128('42')]}}`, this won't work in the operator context.

`mongoize` currently doesn't have context knowledge at all. There is probably a way to communicate some contexts to it but Mongoid does generally permit users to provide MQL in a bunch of places and I'm thinking those might be impossible to change programmatically short of doing much more rewriting of MQL than Mongoid is presently doing.

Reply:
I think at this point what we need is migration guidance and documentation for how to move from string to bigdecimal field storage. The migration query above is excellent and can go into the migration guidance. Additionally it would be good to have a similar snippet for how to query on individual fields with type conversion into bigdecimal, e.g. if I want to find all documents where the value is > 0 with both types how I would do that.

Reply:
Okay! I can't put that all into the documentation PR.... That leaves us with no further code changes then.

Reply:
This test then can be changed to assert that the document is *not* found when its type in the database is different from how Mongoid is presently configured, for our future reference.